ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
+  // cl_dbk_mode_properties has a property not listed in the description of the
+  // defined built-in kernel.
+} else
+  kernel_descriptions.push_back(matmul_desc);
+
+...
+
+cl_program dbk_lib = clCreateProgramWithDefinedBuiltInKernels(
+  context, kernel_descriptions.size(), kernel_descriptors.data(), err);
+
+...
+
+cl_kernel matmul_kernel = clCreateDefinedBuiltinKernel(
+  dkb_lib, matmul_desc, err);
+
+// TBD: allocate space for tensors. Perhaps like cl_qcom_ml_ops: query
+// tensor sizes after the final program has been created or after
+// command buffer (with DBKs within) is finalized. Implementation
+// determines the optimal data layout (opaque to the application) for
+// the tensors based on their usage.  Application uses the tensor
+// sizes to create cl_mem buffers which are bound to the tensors.
+cl_tensor_t lhs_tensor = TBD;
+cl_tensor_t rhs_tensor = TBD;
+cl_tensor_t res_tensor = TBD;
+
+// Transfer data to input tensors
+
+clSetKernelArg(matmul_kernel, 0, sizeof(cl_tensor_t), &lhs_tensor);
+clSetKernelArg(matmul_kernel, 1, sizeof(cl_tensor_t), &rhs_tensor);
+clSetKernelArg(matmul_kernel, 2, sizeof(cl_tensor_t), &res_tensor);
+
+clEnqueueNDRangeKernel(cmd_q, matmul_kernel, 0, NULL, NULL, NULL, 0, NULL, NULL);
+----
 
 === Open questions
 
-. Should we enable launching BiKs from the device side without requiring device-side enqueue? The main problem is those with NDRange as they are not simple single-WI helper functions.
+. Should we enable launching DBKs from the device side without requiring device-side enqueue? The main problem is those with NDRange as they are not simple single-WI helper functions.
 +
 --
 *UNRESOLVED*
 
 --
 
-. Should the NDRange be used at all in BiKs? It feels sort of unnatural as typically the NDRange is used to imply SPMD parallelism while the hardware/firmware is free to choose whatever parallelism degree to implement the function. On the other hand, similar applies to software kernel launches as the work-items can be executed serially if adhering to barrier semantics.
+. Should the NDRange be used at all in DBKs? It feels sort of unnatural as typically the NDRange is used to imply SPMD parallelism while the hardware/firmware is free to choose whatever parallelism degree to implement the function. On the other hand, similar applies to software kernel launches as the work-items can be executed serially if adhering to barrier semantics.
 +
 --
 *UNRESOLVED*
 
 --
 
-. Different accelerators prefer different channel orders (NHWC vs. NCHW...) for the processed data. Should the channel order be passed as a BiK argument (like in the example GEMM's row/column order) or is it better to have different BiK variations for each?
+. Different accelerators prefer different channel orders (NHWC vs. NCHW...) for the processed data. Should the channel order be passed as a DBK argument (like in the example GEMM's row/column order) or is it better to have different DBK variations for each?
```

Comment:
Shouldn't this be abstracted away by the cl_tensor, like we discussed?

Reply:
TBD later.